### ISSUE TITLE
[dcl.init.string] Clarify syntactic forms of initializers that are considered as string literals

### ISSUE THREAD

**Issue by frederick-vs-ja** [cplusplus/CWG#661]
Full name of submitter (unless configured in github; will be published with the issue): Jiang An

Reference (section label): [dcl.init.general], [dcl.init.string]

[dcl.init.general] p16.3:
> If the destination type is an array of characters, an array of `char8_t`, an array of `char16_t`, an array of `char32_t`, or an array of `wchar_t`, and the initializer is a _string-literal_, see [dcl.init.string].

[dcl.init.string] p1
> An array of ordinary character type ([basic.fundamental]), `char8_t` array, `char16_t` array, `char32_t` array, or `wchar_t` array may be initialized by an ordinary string literal, UTF-8 string literal, UTF-16 string literal, UTF-32 string literal, or wide string literal, respectively, or by an appropriately-typed _string-literal_ enclosed in braces ([lex.string]). [...]

The syntactic forms of such initializers are not very clear and there's implementation divergence.

"Good" forms:
- `char a[] = "foo";` - accepted by all known implementations, and clarified to be valid in notes;
- `char a[]("foo");` - accepted by all known implementations;
- `char a[]{"foo"};` - accepted by all known (C++11 and later) implementations;
- `char a[] = {"foo"};` - accepted by all known (C++11 and later) implementations.

Controversial forms with additonal parentheses:
- `char a[] = ("foo");` - diagnosed by GCC (and Comeau per llvm/llvm-project#8418); silently accepted by other implementations;
- `char a[](("foo"));` -  likely same as above;
- `char a[]{("foo")};` - likely same as above;
- `char a[] = {("foo")};` - likely same as above.

Weird forms with additional braces:
- `char arr[]{{"foo"}};` - accepted by MSVC and rejected by others;
- `char arr[] = {{"foo"}};` - same as above.

It seems that we should clarify the permitted syntactic forms, which should include "good" and controversial forms and exclude weird forms.

Link to reflector thread (if any):

Issue description:

Suggested resolution:

**Comment by jensmaurer** [comment]
The rules seem pretty clear to me; [dcl.init.general] p16.3 refers to _string-literal_ (grammar non-terminal), which leaves no room for parentheses.  Similarly, [dcl.init.list] p3.3 handles the brace-enclosed case, also referring to the grammar non-terminal _string-literal_. I'm not seeing how this could be read as allowing parentheses.

**Comment by shafik** [comment]
@jensmaurer I do agree, it looks like it is not allowed, but it does seem like most of the implementations accept the "controversial" forms as an extension. It does indeed look like it is used in the wild:

https://sourcegraph.com/search?q=context:global+lang:c%2B%2B+/%5C%5B%5C%5D%5Cs%2B%5C%3D%5Cs%2B%5C%28%22.*%22%5C%29/+count:11000&patternType=keyword&sm=0

I am not coming up with much background on where this extension came from though.

**Comment by jensmaurer** [comment]
@shafik , half of those hits (approximately) appear to be the same (cloned) code from the clang test suite.

**Comment by Halalaluyafail3** [comment]
> The rules seem pretty clear to me; [dcl.init.general] p16.3 refers to _string-literal_ (grammar non-terminal), which leaves no room for parentheses. Similarly, [dcl.init.list] p3.3 handles the brace-enclosed case, also referring to the grammar non-terminal _string-literal_. I'm not seeing how this could be read as allowing parentheses.

Shouldn't the wording in [expr.prim.paren] apply here?
> A parenthesized expression (E) is a primary expression whose type, result, and value category are identical to those of E. The parenthesized expression can be used in exactly the same contexts as those where E can be used, and with the same meaning, except as otherwise indicated.

Here is an example of this being applied elsewhere:
> A null pointer constant is an integer literal ([lex.icon]) with value zero or a prvalue of type std​::​nullptr_t.

All compilers I know of accept code like `void*p=(0);` even though this paragraph references the term "integer literal" and `(0)` is not a single integer literal token.

C23 also has similar text which states that a parenthesized expression has the same semantics as the expression which it parenthesizes (see 6.5.2p6). Also as I discovered earlier while making a [bug report to GCC](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=115566) about this, GCC accepts using `_Generic` in this context even though it doesn't accept parentheses.

**Comment by frederick-vs-ja** [comment]
> The rules seem pretty clear to me; [dcl.init.general] p16.3 refers to _string-literal_ (grammar non-terminal), which leaves no room for parentheses. Similarly, [dcl.init.list] p3.3 handles the brace-enclosed case, also referring to the grammar non-terminal _string-literal_. I'm not seeing how this could be read as allowing parentheses.

Thanks. But it seems that when the grammar non-terminal _initializer_ corresponds to `("foo")` or `= "foo"`, it's not clear that only `"foo"` is considered to be the initializer in [dcl.init.general] p16.3. Other bullets in [dcl.init.general] p16 seemingly consider outermost enclosing `(` `)` (for direct-initialization) and leading `=` (for copy-initialization) as parts of the initializer.

**Comment by jensmaurer** [comment]
Right. We had wording issues in that area in the past. So, p16.3 has a genuine bug as far as an _initializer_ can never be a _string-literal_, grammatically.

This argument does not apply for a race-enclosed _string-literal_, though.

**Comment by jensmaurer** [comment]
https://godbolt.org/z/EPTrn1PdG

**Comment by jensmaurer** [comment]
CWG2977

**Comment by t3nsor** [comment]
Not in this issue, I guess, but it would be nice if we could get rid of the second sentence of [expr.prim.paren]/1 and just be explicit everywhere about extra parentheses being allowed. I don't think there can be that many places.

**Comment by Halalaluyafail3** [comment]
> It would be good if we could align the C++ answer with the C answer, I think.

[C N3722](https://www.open-std.org/jtc1/sc22/wg14/www/docs/n3722.pdf) was accepted, which intends to allow parentheses in this context.
